lib/Sema/CSFix.cpp

```diff
@@ -1508,3 +1508,38 @@ bool SpecifyKeyPathRootType::diagnose(const Solution &solution,
   
   return failure.diagnose(asNote);
 }
+
+bool UnwrapOptionalBaseKeyPathApplication::diagnose(const Solution &solution,
+                                                    bool asNote) const {
+  MissingOptionalUnwrapKeyPathFailure failure(solution, getFromType(),
+                                              getToType(), getLocator());
+  return failure.diagnose(asNote);
+}
+
+UnwrapOptionalBaseKeyPathApplication *
+UnwrapOptionalBaseKeyPathApplication::attempt(ConstraintSystem &cs, Type baseTy,
+                                              Type rootTy,
+                                              ConstraintLocator *locator) {
+  if(baseTy->hasTypeVariable() || rootTy->hasTypeVariable())
+    return nullptr;
+  
+  // Only diagnose this if base is an optional type.
+  if (!baseTy->getOptionalObjectType())
+    return nullptr;
+  
+  // Ensure we only have one level of optionality so we can
+  // safely suggest unwrapping.
+  auto nonOptional = baseTy->lookThroughSingleOptionalType();
+  if (nonOptional->getOptionalObjectType())
+    return nullptr;
```

Comment:
You can simplify this a bit:

```suggestion
auto nonOptionalTy = baseTy->getOptionalObjectType();
if (!nonOptionalTy || nonOptionalTy-> getOptionalObjectType())
  return nullptr;
```

Reply:
Humm, nice!

Reply:
Fixed, thanks